shims/spark310/src/main/scala/org/apache/spark/sql/rapids/shims/spark310/GpuInMemoryTableScanExec.scala

```diff
@@ -0,0 +1,153 @@
+      }
+    }
+
+    private lazy val inputRDD: RDD[InternalRow] = {
+      if (enableAccumulatorsForTest) {
+        readPartitions.setValue(0)
+        readBatches.setValue(0)
+      }
+
+      val numOutputRows = longMetric("numOutputRows")
+      // Using these variables here to avoid serialization of entire objects (if referenced
+      // directly) within the map Partitions closure.
+      val relOutput = relation.output
+      val serializer = relation.cacheBuilder.serializer
+
+      // update SQL metrics
+      val withMetrics =
+        filteredCachedBatches().map{ batch =>
+          if (enableAccumulatorsForTest) {
+            readBatches.add(1)
+          }
+          numOutputRows += batch.numRows
+          batch
+        }
+      val rows = serializer.convertCachedBatchToInternalRow(withMetrics, relOutput, attributes,
+        conf)
+      if (enableAccumulatorsForTest) {
+        def incParts(index: Int, iter: Iterator[InternalRow]): Iterator[InternalRow] = {
+          if (iter.hasNext) {
+            readPartitions.add(1)
+          }
+          iter
+        }
+        rows.mapPartitionsWithIndexInternal(incParts)
+      } else {
+        rows
+      }
+    }
+
+    override def output: Seq[Attribute] = attributes
+
+    private def updateAttribute(expr: Expression): Expression = {
+      // attributes can be pruned so using relation's output.
+      // E.g., relation.output is [id, item] but this scan's output can be [item] only.
+      val attrMap = AttributeMap(relation.cachedPlan.output.zip(relation.output))
+      expr.transform {
+        case attr: Attribute => attrMap.getOrElse(attr, attr)
+      }
+    }
+
+    // The cached version does not change the outputPartitioning of the original SparkPlan.
+    // But the cached version could alias output, so we need to replace output.
+    override def outputPartitioning: Partitioning = {
+      relation.cachedPlan.outputPartitioning match {
+        case e: Expression => updateAttribute(e).asInstanceOf[Partitioning]
+        case other => other
+      }
+    }
+
+    // The cached version does not change the outputOrdering of the original SparkPlan.
+    // But the cached version could alias output, so we need to replace output.
+    override def outputOrdering: Seq[SortOrder] =
+      relation.cachedPlan.outputOrdering.map(updateAttribute(_).asInstanceOf[SortOrder])
+
+    lazy val enableAccumulatorsForTest: Boolean = sqlContext.conf.inMemoryTableScanStatisticsEnabled
+
+    // Accumulators used for testing purposes
+    lazy val readPartitions = sparkContext.longAccumulator
+    lazy val readBatches = sparkContext.longAccumulator
+
+    private val inMemoryPartitionPruningEnabled = sqlContext.conf.inMemoryPartitionPruning
+
+    private def filteredCachedBatches() = {
+      // Right now just return the batch without filtering
+      relation.cacheBuilder.cachedColumnBuffers
+    }
+
+    protected override def doExecute(): RDD[InternalRow] = {
+      inputRDD
```

Comment:
This should probably blow up because we are only going to replace this if the output is columnar on the GPU.  This means we can delete all of the code around inputRDD too.

Reply:
So if this going to blow up, the entire code path that calls into convertCachedBatchToInternalRow will be dead code?